pkgs/build-support/rust/fetch-cargo-vendor-util.py

```diff
@@ -0,0 +1,247 @@
+            git_sha_rev = source_info["git_sha_rev"]
+
+            git_tree = os.path.join(vendor_staging_dir, "git", git_sha_rev)
+            crate_manifest_path = find_crate_manifest_in_tree(git_tree, pkg["name"])
+            crate_tree = os.path.dirname(crate_manifest_path)
+
+            dir_name = f"{pkg["name"]}-{pkg["version"]}"
+            crate_out_dir = os.path.join(out_dir, dir_name)
+            shutil.copytree(crate_tree, crate_out_dir)
+            os.chmod(crate_out_dir, 0o755)
+
+            with open(crate_manifest_path, "r") as f:
+                manifest_data = f.read()
+
+            if "workspace" in manifest_data:
+                workspace_root = get_manifest_metadata(crate_manifest_path)["workspace_root"]
+                os.chmod(os.path.join(crate_out_dir, "Cargo.toml"), 0o644)
+                subprocess.check_output(["replace-workspace-values", os.path.join(crate_out_dir, "Cargo.toml"), os.path.join(workspace_root, "Cargo.toml")])
+
+            with open(os.path.join(crate_out_dir, ".cargo-checksum.json"), "w") as f:
+                f.write('{"files":{},"package":null}')
+
+            eprint(crate_out_dir)
+
+            source_key = source[0:source.find("#")]
+
+            if source_key not in seen_source_keys:
+                config_lines.append(f'[source."{source_key}"]')
+                config_lines.append(f'git = "{source_info["url"]}"')
+                if source_info["type"] is not None:
+                    config_lines.append(f'{source_info["type"]} = "{source_info["value"]}"')
+                config_lines.append('replace-with = "vendored-sources"')
+
+                seen_source_keys.add(source_key)
+
+        elif source.startswith("registry+"):
+            filename = f"{pkg["name"]}-{pkg["version"]}.tar.gz"
+            dir_name = f"{pkg["name"]}-{pkg["version"]}"
+            crate_out_dir = os.path.join(out_dir, dir_name)
+            tarball_path = os.path.join(vendor_staging_dir, "tarballs", filename)
+
+            os.mkdir(crate_out_dir)
+            subprocess.check_output(["tar", "xf", tarball_path, "-C", crate_out_dir, "--strip-components=1"])
+
+            with open(os.path.join(crate_out_dir, ".cargo-checksum.json"), "w") as f:
+                f.write(f'{{"files":{{}},"package":"{pkg["checksum"]}"}}')
+
+            eprint(crate_out_dir)
+
+        else:
+            raise Exception(f"Can't process source: {source}.")
+
+    os.mkdir(os.path.join(out_dir, ".cargo"))
+    with open(os.path.join(out_dir, ".cargo", "config.toml"), "w") as config_file:
+        config_file.writelines(line + "\n" for line in config_lines)
+
+
+def main() -> None:
+    method = sys.argv[1]
+
+    if method == "create-vendor-staging":
+        create_vendor_staging(lockfile_path=sys.argv[2], out_dir=sys.argv[3])
+    elif method == "create-vendor":
+        create_vendor(vendor_staging_dir=sys.argv[2], out_dir=sys.argv[3])
+    else:
+        raise Exception(f"Unknown method: {method}")
```

Comment:
I would make create your own `Exception` at the start of the script and than catch this exception in the main method:

```
try:
    ...
exception Error as e:
   eprint(e)
```

This won't show than error stacktraces for our own errors, which is a bit more user friendly (because users often don't find the relevant error from a stack trace).